src/hotspot/cpu/ppc/stubGenerator_ppc.cpp

```diff
@@ -4016,6 +4016,413 @@ class StubGenerator: public StubCodeGenerator {
+// at each location, all values in expanded are compared to 31.  Using
+// vsel, values higher than 31 use the results from the upper 32 bytes of
+// the lookup operation, while values less than or equal to 31 use the
+// lower 32 bytes of the lookup operation.  Power10 and beyond can save the
+// compare instruction, because the comparison is done within xxpermx
+// itself. TODO: use xxpermx,xxpermx,vor on P10 when instruction prefixes are
+// available in assembler_ppc.*
+
+#define ENCODE_CORE                                                        \
+    __ xxperm(input->to_vsr(), input->to_vsr(), expand_permute);           \
+    __ vsrb(rshift, input, expand_rshift);                                 \
+    __ vand(rshift, rshift, expand_rshift_mask);                           \
+    __ vslo(lshift, input, vec_8s);                                        \
+    __ vslb(lshift, lshift, expand_lshift);                                \
+    __ vand(lshift, lshift, expand_lshift_mask);                           \
+    __ vor(expanded, rshift, lshift);                                      \
+    __ vperm(encoded_00_31, vec_base64_00_15, vec_base64_16_31, expanded); \
+    __ vperm(encoded_32_63, vec_base64_32_47, vec_base64_48_63, expanded); \
+    __ vcmpgtub(gt_31, expanded, vec_31s);                                 \
+    __ vsel(expanded, encoded_00_31, encoded_32_63, gt_31);
+
+// Intrinsic function prototype in Base64.java:
+// private void encodeBlock(byte[] src, int sp, int sl, byte[] dst, int dp, boolean isURL) {
+
+  address generate_base64_encodeBlock() {
+    __ align(CodeEntryAlignment);
+    StubCodeMark mark(this, "StubRoutines", "base64_encodeBlock");
+    address start   = __ function_entry();
+
+    static const __vector unsigned char expand_permute_val = {
+      ARRAY_TO_LXV_ORDER(
+      0,  4,  5,  6,
+      0,  7,  8,  9,
+      0, 10, 11, 12,
+      0, 13, 14, 15 ) };
+
+    static const __vector unsigned char expand_rshift_val = {
+      ARRAY_TO_LXV_ORDER(
+      0, 6, 4, 2,
+      0, 6, 4, 2,
+      0, 6, 4, 2,
+      0, 6, 4, 2 ) };
+
+    static const __vector unsigned char expand_rshift_mask_val = {
+      ARRAY_TO_LXV_ORDER(
+      0b00000000, 0b00000011, 0b00001111, 0b00111111,
+      0b00000000, 0b00000011, 0b00001111, 0b00111111,
+      0b00000000, 0b00000011, 0b00001111, 0b00111111,
+      0b00000000, 0b00000011, 0b00001111, 0b00111111 ) };
+
+    static const __vector unsigned char expand_lshift_val = {
+      ARRAY_TO_LXV_ORDER(
+      0, 2, 4, 0,
+      0, 2, 4, 0,
+      0, 2, 4, 0,
+      0, 2, 4, 0 ) };
+
+    static const __vector unsigned char expand_lshift_mask_val = {
+      ARRAY_TO_LXV_ORDER(
+      0b00111111, 0b00111100, 0b00110000, 0b00000000,
+      0b00111111, 0b00111100, 0b00110000, 0b00000000,
+      0b00111111, 0b00111100, 0b00110000, 0b00000000,
+      0b00111111, 0b00111100, 0b00110000, 0b00000000 ) };
+
+    static const __vector unsigned char base64_00_15 = {
```

Comment:
I don't really like so many individual arrays of constants. They could get put into one larger block of memory and accessed by lvx with offset.

Reply:
Interesting idea!  The point there would be to save the initialization time of loading the address constants?  Otherwise, it has a negative affect on code readability.

Reply:
Right, we could use much shorter code for loading the constants. I don't think readability would suffer if you only put constants of the same kind (like the base64_... ones) into single arrays. And even if you put all constants into one array I guess it could be made readable by good comments. But there may be a trade-off.

Reply:
Ok, I will definitely give this a go.  If it looks maintainable, I'll do it.  The loads aren't inside the loop, but still it would be nice especially for calls to the encoder that have a short buffer.

Reply:
I have pushed changes for this issue.  Take a look and see if you agree that what I did is acceptable, especially with respect to the hard-coded offsets.  I'd be interested to hear If you have a better idea about to maintain them.

Reply:
It'd be possible to put them into a struct and have the offsets computed by the C++ compiler, but I'm fine with your version. I guess there won't be many manual changes any more.
I'll test the code a bit longer, but it looks good so far.

Reply:
Oh, that's a great idea!  I'm willing to make that change, if you don't mind one last cycle of review.

Reply:
It's your choice. I'm also fine with your current version.
Btw. I had also thought a bit about the division by constants. It's only used outside the loop. Otherwise I'd have suggested to use mulhd (e.g. positive long can get divided by 3 by mulhd with 0x5555555555555556, division by 12 by mulhd with 0x2aaaaaaaaaaaaaab + right shift by one). Such fast code gets generated by C2 compiler.

Reply:
> Btw. I had also thought a bit about the division by constants. It's only used outside the loop. Otherwise I'd have suggested to use mulhd (e.g. positive long can get divided by 3 by mulhd with 0x5555555555555556, division by 12 by mulhd with 0x2aaaaaaaaaaaaaab + right shift by one). Such fast code gets generated by C2 compiler.

That's very interesting.  I'm not clear how that math works, but I tried it out in C using  the__int128 type, and it works as advertised.  However, since the division takes place outside the loop, it's hard to justify putting something this obtuse in the code.  Thanks for the thought, though.  I'll do some google searches to find out why this works for my own education.

I will put the struct change in tomorrow.  I think that's much cleaner.

Reply:
Ah, I see now.  It's basically multiplying by the binary-point reciprocal, and then accounting for the number of number of binary point digits, which in this case effectively means take the answer from the top 64 bits of the 128-bit result.

Reply:
Right. It's tricky to make sure there will be no rounding errors, so the algorithm to determine the reciprocal together with the shift amount is not trivial. It's implemented in "magic_int_divide_constants" (divnode.cpp) and taken from "Hacker's Delight by Henry S. Warren, Jr." where you can find a detailed explanation. But we warned, verifying the proof takes time!
However, you can easily compute the constants by compiling a Java long division by C2 and disassembling.

Reply:
I thought there might be some trickiness in that constant because the last digit was rounded.   Thanks for the pointers!